### PR TITLE
Training for Corrdiff-mini: a simple CorrDiff for testing and education

### PR DESCRIPTION
**Pull request by jleinonen**
# Modulus Pull Request

## Description
This PR adds training support for CorrDiff-mini, a simple CorrDiff with a compact dataset and a smaller network that can be used for education and testing.

Specifically, it includes the following additions:
- Dataset implementation for the compact HRRR dataset in `datasets/hrrrmini.py`
- Config files to define the CorrDiff-Mini dataset, networks and training.
- Bug fixes and feature enhancements to allow the implementation, such as overriding network parameters from a config file.
- More consistent handling of periodic tasks during training.

CorrDiff-mini generation will be added in a later PR.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: examples/generative/corrdiff/train.py [hunk 1]
```diff
@@ -303,57 +327,46 @@ def main(cfg: DictConfig) -> None:
...
-                f"peak_gpu_mem_reserved_gb {(torch.cuda.max_memory_reserved(dist.device) / 2**30):<6.2f}"
-            ]
-            logger0.info(" ".join(fields))
-        torch.cuda.reset_peak_memory_stats()
+        if is_time_for_periodic_task(
+            cfg.training.io.print_progress_freq, rank_0_only=True
+        ):
+            # Print stats if we crossed the printing threshold with this batch
+            tick_end_time = time.time()
+            if dist.rank == 0:
```
Comment (mnabian) [inline examples/generative/corrdiff/train.py:335]: I don't think this is needed as you have already set `rank_0_only=True`

File: examples/generative/corrdiff/train.py [hunk 2]
```diff
@@ -207,19 +212,34 @@ def main(cfg: DictConfig) -> None:
...
+        cur_nimg = load_checkpoint(
+            path=f"checkpoints_{cfg.model.name}",
+            models=model,
+            optimizer=optimizer,
+            device=dist.device,
+        )
+    except:
+        cur_nimg = 0
+
+    def is_time_for_periodic_task(freq, rank_0_only=False):
```
Comment (mnabian) [inline examples/generative/corrdiff/train.py:225]: Maybe consider moving this to `training_helpers.py`

File: examples/generative/corrdiff/datasets/hrrrmini.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    mean = np.array([stats[group][v]["mean"] for v in variables])[:, None, None].astype(
+        np.float32
+    )
+    std = np.array([stats[group][v]["std"] for v in variables])[:, None, None].astype(
+        np.float32
+    )
+    return (mean, std)
+
+
+def _convert_datetime_to_cftime(
```
Comment (mnabian) [inline examples/generative/corrdiff/datasets/hrrrmini.py:166]: The same utility exists in `modulus,generative.utils`. Please import from there.

File: examples/generative/corrdiff/datasets/hrrrmini.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+import cftime
+import numpy as np
+from numba import jit, prange
+import xarray as xr
+
+from .base import ChannelMetadata, DownscalingDataset
+
+
+class HRRRMiniDataset(DownscalingDataset):
```
Comment (mnabian) [inline examples/generative/corrdiff/datasets/hrrrmini.py:30]: Please add API documentation here, as well as a general documentation in the README.

Comment (mnabian) [reply examples/generative/corrdiff/datasets/hrrrmini.py:30]: Also, what do you think about moving this to the Modulus datapipes directory?

**Comment by mnabian** [comment]
Would be great to add documentation for the CorrDiff-mini example

File: examples/generative/corrdiff/datasets/hrrrmini.py (hunk 3 above)
Comment (jleinonen) [reply examples/generative/corrdiff/datasets/hrrrmini.py]: Done in https://github.com/NVIDIA/modulus/pull/636/commits/3df9d26f94ec5d13075e1bd6b1f919a9b07b114c.

File: examples/generative/corrdiff/train.py (hunk 2 above)
Comment (jleinonen) [reply examples/generative/corrdiff/train.py]: Done in https://github.com/NVIDIA/modulus/pull/636/commits/3df9d26f94ec5d13075e1bd6b1f919a9b07b114c.

File: examples/generative/corrdiff/train.py (hunk 1 above)
Comment (jleinonen) [reply examples/generative/corrdiff/train.py]: Removed in https://github.com/NVIDIA/modulus/pull/636/commits/3df9d26f94ec5d13075e1bd6b1f919a9b07b114c.

File: examples/generative/corrdiff/datasets/hrrrmini.py (hunk 4 above)
Comment (jleinonen) [reply examples/generative/corrdiff/datasets/hrrrmini.py:31]: I would suggest keeping it as is for now - I'd support moving the`DownscalingDataset` base class to Modulus though, once the API becomes stable. At least the lat/lon handling should be figured out before I'd consider it stable.

**Review by mnabian — APPROVED** [review]

**Pull request merged**